docs/proposals/vip-4.md

```diff
@@ -0,0 +1,217 @@
+|                        `preStoreVersionSwap` |        ✅         |         ✅         |    ✅     |
+|                       `postStoreVersionSwap` |        ✅         |         ✅         |    ✅     |
+|                            `preEndOfPushJob` |        ✅         |         ✅         |    ✅     |
+|                           `postEndOfPushJob` |        ✅         |         ✅         |    ✅     |
+
+The cells with numbers in them represent special cases, which are explained below:
+
+1️⃣ : Schema registration hooks in push jobs could only be invoked in cases where auto-registration is enabled and the 
+new schema originates from the push job itself, whereas schema registrations which are performed directly on the 
+controller could not trigger the hook.
+
+2️⃣ : The hook for the start of ingestion for Da Vinci Clients is tricky for a few reasons. The start of ingestion is 
+controlled by updating the Meta Store, which is a non-replicated system store updated by child controllers, so those
+must be involved (either by running the hook there in the first place, or by having some mechanism that enables the
+parent or VPJ to inform the child controllers of when the system store is eligible for getting updated, such as by
+adding a new field to the `AddVersion` admin channel command). However, see Rollback Support below for why running hooks
+in the child controller may be insufficient.
+
+### Rollback Support
+
+In order to support the ability for the hooks which return the `StoreVersionLifecycleEventOutcome` to rollback, the most
+natural way to achieve this is likely to involve the parent controller, either by having those hooks run there in the 
+first place, or by having a propagation mechanism to it:
+
+* If the hooks are executed in the child controller, the propagation mechanism might be to extend the job status check 
+  which the parent controller does periodically in order for the child to inform the parent of the need to rollback, or 
+  else build a new mechanism for the child to directly interact with the parent (or even with other child controllers 
+  directly...).
+
+* If the hooks are executed in VPJ, then it would need to interact with the parent via the controller client to trigger 
+  the rollback.
+
+### Configs
+
+There needs to be new configs:
+
+* `venice.store.lifecycle.hooks`: Comma-separated list of FQCN of the hook implementations to load.
+
+* `venice.store.lifecycle.hooks.threads`: Number of threads used by the hooks framework to execute all hooks.
+
+* `venice.store.lifecycle.hooks.wait.interval.seconds`: The time to wait before re-invoking a hook which returned `WAIT` 
+  (default: 60 seconds).
+
+* `venice.store.lifecycle.hooks.configs.<arbitrary>`: Any number of configs to be passed (after clipping everything
+  before the `<arbitrary>` part) into the hooks constructor.
+
+In addition, the store config will get a new `Map<String, String>` of store-level config overrides.
```

Comment:
Why do we need this ? these can be overriden from pushjob itself?

Reply:
Good question!

This is not indicated in this doc... but we have some internal requirement to control in a centralized manner some of the hooks' behaviors. This will be easier to accomplish via store configs than push job configs.

For behaviors that we wish to allow the user to override arbitrarily, it is conceivable to have those as VPJ configs...